File: python/modcdp/injector/DiscoveredExtensionInjector.py
from __future__ import annotations

from typing import Any

from ..injector.ExtensionInjector import DEFAULT_SERVICE_WORKER_PROBE_TIMEOUT_MS, DEFAULT_SERVICE_WORKER_READY_TIMEOUT_MS, ExtensionInjector, ExtensionInjectionResult


def _timeout(value: Any, fallback: int) -> int:
    return fallback if value is None else int(value)


class DiscoveredExtensionInjector(ExtensionInjector):
    def inject(self) -> ExtensionInjectionResult | None:
        discovered = self._discoverReadyServiceWorker()
        if discovered:
            return {**discovered, "source": "discovered"}
        if self.options.get("injector_trust_service_worker_target"):
            waited = self._waitForReadyServiceWorker(
                _timeout(self.options.get("injector_service_worker_probe_timeout_ms"), DEFAULT_SERVICE_WORKER_PROBE_TIMEOUT_MS),
                matched_only=True,
            )
            if waited:
                return {**waited, "source": "discovered"}
        if not self.options.get("injector_require_service_worker_target"):
            return None
        waited = self._waitForReadyServiceWorker(
            _timeout(self.options.get("injector_service_worker_ready_timeout_ms"), DEFAULT_SERVICE_WORKER_READY_TIMEOUT_MS),
            matched_only=bool(self.options.get("injector_trust_service_worker_target")),
        )
        if waited:
            return {**waited, "source": "discovered"}
        matchers = ", ".join(
            [
                *(self.options.get("injector_service_worker_url_includes") or []),
                *(self.options.get("injector_service_worker_url_suffixes") or []),
            ]
        )
        raise RuntimeError(f"Required ModCDP service worker target was not visible ({matchers or 'no matcher'}).")

File: go/modcdp/injector/DiscoveredExtensionInjector.go
package injector

import (
	"fmt"
	"strings"
)

type DiscoveredExtensionInjector struct {
	ExtensionInjector
}

func NewDiscoveredExtensionInjector(options ExtensionInjectorConfig) DiscoveredExtensionInjector {
	return DiscoveredExtensionInjector{ExtensionInjector: NewExtensionInjector(options)}
}

func (i *DiscoveredExtensionInjector) Inject() (*ExtensionInjectionResult, error) {
	discovered, err := i.discoverReadyServiceWorker(false)
	if err != nil || discovered != nil {
		if discovered != nil {
			discovered.Source = "discovered"
		}
		return discovered, err
	}
	if i.Options.InjectorTrustServiceWorkerTarget {
		waited, err := i.waitForReadyServiceWorker(i.Options.InjectorServiceWorkerProbeTimeoutMS, true)
		if err != nil || waited != nil {
			if waited != nil {
				waited.Source = "discovered"
			}
			return waited, err
		}
	}
	if !i.Options.InjectorRequireServiceWorkerTarget {
		return nil, nil
	}
	waited, err := i.waitForReadyServiceWorker(i.Options.InjectorServiceWorkerReadyTimeoutMS, i.Options.InjectorTrustServiceWorkerTarget)
	if err != nil || waited != nil {
		if waited != nil {
			waited.Source = "discovered"
		}
		return waited, err
	}
	matchers := append(append([]string{}, i.Options.InjectorServiceWorkerURLIncludes...), i.Options.InjectorServiceWorkerURLSuffixes...)
	matcherText := strings.Join(matchers, ", ")
	if matcherText == "" {
		matcherText = "no matcher"
	}
	return nil, fmt.Errorf("required ModCDP service worker target was not visible (%s)", matcherText)
}
